pkg/runtime.v2/framework/plugins/torch/torch.go

```diff
@@ -43,11 +48,89 @@ func (t *Torch) Name() string {
+func (t *Torch) EnforceMLPolicy(info *runtime.Info, trainJob *kubeflowv2.TrainJob, runtimeMLPolicy *kubeflowv2.MLPolicy) error {
+	if info == nil || runtimeMLPolicy == nil || runtimeMLPolicy.Torch == nil {
 		return nil
 	}
-	// TODO: Need to implement main logic.
+
+	// TrainJob contains the actual information for the Trainer.
+	numNodes := runtimeMLPolicy.NumNodes
+	if trainJob.Spec.Trainer != nil && trainJob.Spec.Trainer.NumNodes != nil {
+		numNodes = trainJob.Spec.Trainer.NumNodes
+	}
+	info.Trainer.NumNodes = numNodes
+
+	numProcPerNode := runtimeMLPolicy.Torch.NumProcPerNode
+	if trainJob.Spec.Trainer != nil && trainJob.Spec.Trainer.NumProcPerNode != nil {
+		numProcPerNode = trainJob.Spec.Trainer.NumProcPerNode
+	}
+
+	// Update envs for Info object.
+	// Add PyTorch distributed "PET_" values for torchrun
+	// TODO (andreyvelich): Add validation to check that TrainJob doesn't have "PET_" envs.
+	// TODO (andreyvelich): We should validate that envs from different plugins don't conflict with each other.
+	// Ref: https://github.com/kubeflow/training-operator/pull/2308#discussion_r1823229940
+	infoEnvs := []corev1.EnvVar{
+		{
+			Name:  constants.TorchEnvNumNodes,
+			Value: fmt.Sprintf("%d", ptr.Deref(numNodes, 1)),
+		},
+		{
+			Name:  constants.TorchEnvNumProcPerNode,
+			Value: ptr.Deref(numProcPerNode, "auto"),
+		},
+		{
+			Name: constants.TorchEnvNodeRank,
+			ValueFrom: &corev1.EnvVarSource{
+				FieldRef: &corev1.ObjectFieldSelector{
+					FieldPath: constants.JobCompletionIndexFieldPath,
+				},
+			},
+		},
+		{
+			Name:  constants.TorchEnvMasterAddr,
+			Value: fmt.Sprintf("%s-%s-0-0.%s", trainJob.Name, constants.JobTrainerNode, trainJob.Name),
+		},
+		{
+			Name:  constants.TorchEnvMasterPort,
+			Value: fmt.Sprintf("%d", constants.ContainerTrainerPort),
+		},
+	}
+
+	// Set for all Info envs.
+	envNames := sets.New[string]()
+	for _, env := range infoEnvs {
+		envNames.Insert(env.Name)
+	}
+	// Info envs take precedence over TrainJob envs.
+	if trainJob.Spec.Trainer != nil {
+		for _, env := range trainJob.Spec.Trainer.Env {
+			if !envNames.Has(env.Name) {
+				info.Trainer.Env = append(info.Trainer.Env, corev1.EnvVar{Name: env.Name, Value: env.Value})
+			}
+		}
+	}
+	// Insert Torch distributed envs into the list end.
+	info.Trainer.Env = append(info.Trainer.Env, infoEnvs...)
+
+	// Add container port for the headless service.
+	info.Trainer.ContainerPort = &corev1.ContainerPort{
+		ContainerPort: constants.ContainerTrainerPort,
+	}
+
+	// Update total Pod requests for the PodGroupPolicy plugin.
+	for rName := range info.TotalRequests {
+		// For other Jobs like the Initializer, replica is always equal to 1.
+		// TODO (andreyvelich): Add support for total requests from the TrainJob's ResourcesPerNode.
+		if rName == constants.JobTrainerNode {
+			info.TotalRequests[rName] = runtime.TotalResourceRequest{
+				Replicas:    ptr.Deref(numNodes, 1),
```

Comment:
nit: we can move this 1 to constants file defaultReplicas

Reply:
Done.